Implement command line handling

marhkb opened the pull request:
This PR is split up into 3 commits.

1. Implement basic command line handling so that we can print the application version number.
2. Setting the log level using a command line parameter.
3. Only for debug builds: Setting the data directory using a command line parameter.

More detailed descriptions can be found in the commit messages.

melix99 commented on src/application.rs:
```diff
@@ -132,10 +137,6 @@ impl Application {
     }
 
     pub fn run(&self) {
```
This function can be removed now.

melix99 commented on src/cli.rs:
```diff
@@ -0,0 +1,37 @@
+use std::convert::TryInto;
```
This use declaration should be part of the group below.

melix99 commented on src/cli.rs:
```diff
@@ -0,0 +1,37 @@
+use gtk::prelude::{ApplicationExt, IsA};
+use gtk::{gio, glib};
+
+pub fn setup<A: IsA<gio::Application>>(app: A) -> A {
```
I guess this can be part of main.rs? Or maybe you plan on adding more things to it and you want to keep it separate?

marhkb replied:
No, I think there's nothing much to add.

melix99 commented on src/cli.rs:
```diff
@@ -0,0 +1,37 @@
+pub fn setup<A: IsA<gio::Application>>(app: A) -> A {
+    app.add_main_option(
+        "version",
+        'v'.try_into().unwrap(),
```
```suggestion
        b'v'.into(),
```

To avoid an unwrap.

melix99 commented on src/cli.rs:
```diff
@@ -0,0 +1,37 @@
+
+    app.add_main_option(
+        "log-level",
+        'l'.try_into().unwrap(),
```
```suggestion
        b'l'.into(),
```

To avoid an unwrap.

melix99 commented on src/cli.rs:
```diff
@@ -0,0 +1,37 @@
+    #[cfg(debug_assertions)]
+    app.add_main_option(
+        "data-dir",
+        'd'.try_into().unwrap(),
```
```suggestion
        b'd'.into(),
```

To avoid an unwrap.

melix99 commented on src/cli.rs:
```diff
@@ -0,0 +1,37 @@
+        'l'.try_into().unwrap(),
+        glib::OptionFlags::NONE,
+        glib::OptionArg::String,
+        &gettext("Specify the minumum log level"),
```
```suggestion
        &gettext("Specify the minimum log level"),
```

melix99 commented on src/cli.rs:
```diff
@@ -0,0 +1,37 @@
+        'd'.try_into().unwrap(),
+        glib::OptionFlags::NONE,
+        glib::OptionArg::String,
+        &gettext("Specify another data directory"),
```
```suggestion
        &gettext("Specify a different data directory"),
```

melix99 commented on src/main.rs:
```diff
@@ -1,28 +1,34 @@
 mod utils;
 mod window;
 
+use std::path::PathBuf;
```
All this group of use declarations should be grouped below the `use self:...` declarations.

marhkb committed:
cli: Implement basic cli handling

For this purpose, external libraries such as 'clap' are intentionally
omitted. On the one hand more complex functions like subcommands are
not needed and on the other hand clap interferes with the gtk specific
command line handling. Instead, command line options are added to the
gio::Application.

In this first commit to handle command line parameters, the version
number of Telegrand can already be printed.

marhkb committed:
cli: Set log level using parameter

At the moment the log level can only be influenced with the environment
variable 'RUST_LOG'. This is very specific for the log implementation
'pretty_env_logger' and would not work anymore if another log implemen-
tation would be used in the future. In addition, this is not very ex-
ploratory for the user.

With this commit, the log level can be set as a command line option. On
the one hand this makes the options more explorative for the user. On
the other hand, the option serves as a kind of interface that can be
used to integrate other log implementations in the future.

If the user does not specify a log level, 'debug' is used by default
for debug builds and 'info' for release builds.
On top of that, the log level is also transmitted to tdlib.

marhkb committed:
cli: Set data dir by parameter for debug builds

Especially for the development of Telegrand it makes sense to specify
another data directory via command line arguments. Because often you
work at different places in the application at the same time (login vs.
inapp) and so you would have the possibility to switch back and forth
quickly.

This option for alternative data directories is only enabled for debug
builds, because it would make little sense to provide them in releases.

melix99 commented on src/main.rs:
```diff
@@ -6,23 +6,28 @@ mod session;
 mod utils;
 mod window;
 
+use crate::config::VERSION;
```
This should go below the use declarations of the external crates (grouped separately from them).

To clarify, my preferred order and grouping for use declarations is:

1. Module imports
2. 'self' use declarations (which are useful to simplify the access to the inner objects of the modules)
3. Use declarations of external crates (including std)
4. Use declarations of the current crate

marhkb replied:
I overlooked that one. Nevertheless, I removed the use statement altogether, since `VERSION` is only used once anyway. This looks cleaner then.
